transport/tchannel/transport.go

```diff
@@ -0,0 +1,211 @@
+package tchannel
+
+import (
+	"sync"
+
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/introspection"
+	intsync "go.uber.org/yarpc/internal/sync"
+	"go.uber.org/yarpc/peer/hostport"
+
+	"github.com/opentracing/opentracing-go"
+	"github.com/uber/tchannel-go"
+)
+
+// Transport is a TChannel transport suitable for use with YARPC's peer
+// selection system.
+// The transport implements peer.Transport so multiple peer.List
+// implementations can retain and release shared peers.
+// The transport implements transport.Transport so it is suitable for lifecycle
+// management.
+type Transport struct {
+	lock sync.Mutex
+	once intsync.LifecycleOnce
+
+	ch     Channel
+	router transport.Router
+	tracer opentracing.Tracer
+	addr   string
+
+	peers map[string]*hostport.Peer
+}
+
+// NewTransport is a YARPC transport that facilitates sending and receiving
+// YARPC requests through TChannel.
+// It uses a shared TChannel Channel for both, incoming and outgoing requests,
+// ensuring reuse of connections and other resources.
+//
+// Either the local service name (with the ServiceName option) or a user-owned
+// TChannel (with the WithChannel option) MUST be specified.
+func NewTransport(opts ...TransportOption) (*Transport, error) {
+	var config transportConfig
+	config.tracer = opentracing.GlobalTracer()
+	for _, opt := range opts {
+		opt(&config)
+	}
+
+	// Attempt to construct a channel on behalf of the caller if none given.
+	// Defer the error until Start since NewChannelTransport does not have
+	// an error return.
+	var err error
+	ch := config.ch
+
+	if ch == nil {
+		if config.name == "" {
+			err = errChannelOrServiceNameIsRequired
+		} else {
+			opts := tchannel.ChannelOptions{Tracer: config.tracer}
+			ch, err = tchannel.NewChannel(config.name, &opts)
+		}
+	}
+
+	return &Transport{
+		ch:     ch,
+		addr:   config.addr,
+		tracer: config.tracer,
+		peers:  make(map[string]*hostport.Peer),
+	}, err
+}
+
+// ListenAddr exposes the listen address of the transport.
+func (t *Transport) ListenAddr() string {
```

Comment:
Yeah, we need to figure out whether we’re going to support multiple listening sockets on a single transport. If we do, this must be removed. Otherwise, we need a way to enumerate the inbounds and get their addresses. This method does need to exist *somewhere* so we can discover the OS assigned port when we listen on :0.